### PR TITLE
Add Support for Exporting Boundaries

### LINKED ISSUE #93: Administrator Role Support in front-end
## Overview

Ensure that an Administrator can act as Validator in all cases. Ensure they can also act as Contributor in some cases (ability to edit submissions, etc).

### LINKED ISSUE #172: NavBar Should Show Submission Utility, Not Selected Utility
## Overview

The NavBar should always show the submission's utility, not the one that the user has selected:

https://user-images.githubusercontent.com/1430060/199102062-3c8b8710-865d-4a90-8ba2-c36adc984b24.mp4

### LINKED ISSUE #184: Automatically determine utility at utillity specific URLs
## Overview

Going straight to a utility specific URL (e.g. `/submissions/4` or `/draw/4`) still shows the utility selector. This can be annoying and potentially confusing. If a user selects a utility not related to that of the URL's item, then the user will still see the URL item, but will have a different utility selected.

### Describe the solution you'd like

Automatically set the utility when the user visits one of these URLs

We should wrap all of the utility specific routes in a component which defers rendering of its children until the utility is set. This would probably be easier to do if we moved the draw page "into" submissions: `/submissions/4/draw`.

### LINKED ISSUE #211: Hide Add Map Button for Validators
## Overview

If the logged in user is a Validator, the Add Map button should be hidden:

image

### LINKED ISSUE #216: Add Ability to Export Boundaries
## Overview

Add a button in the UI for administrators that hits an endpoint only they can access, that exports all the approved boundaries with their shapes as a geojson.

Since the amount of data can potentially be very large, design a query that generates the geojson in the database and directly send / stream that data to the user, rather than constructing it in Django.

### PR DESCRIPTION
**Pull request by rajadain**
## Overview

Adds two new endpoints to export boundaries: one for only approved boundaries, and one for all boundaries. Adds a button in the front-end for administrators to do so.

Also cleans up some minor issues.

Closes #216 
Closes #172 
Closes #93 
Closes #211 
Closes #184 

### Demo

image

image

### Notes

5e6fbf2d8f6c7318bceff46209fbf3d1f3c34289 was done when I thought that the button would live in the List NavBar, but in the end it made more sense to place it on the main page itself. It can be excluded, but I think it is still valuable, so kept it in.

## Testing Instructions

- Checkout this branch and `server`
- Login as an Administrator
- Go to the Submissions List
  - [x] Ensure you see a button to export boundaries
  - [x] Ensure you are able to export all boundaries
  - [x] Ensure you are able to export approved boundaries
- Log out, and login as a Contributor
- Open this URL: http://localhost:4545/submissions/1
- ~~Select "Other Utility" in the selector dropdown~~
  - [ ] ~~Ensure the loaded page says "Azavea Test Utility" in the top-left, not "Other Utility"~~
  - [x] Ensure you do not see the utility selector dropdown
  - [x] Ensure the loaded page says "Azavea Test Utility" in the top-left
- Open this URL: http://localhost:4545/submissions
  - [x] Ensure you do see the utility selector dropdown
  - [x] Ensure you are taken to the appropriate submissions list after

## Checklist

- [ ] `fixup!` commits have been squashed
- [ ] `CHANGELOG.md` updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [ ] `README.md` updated if necessary to reflect the changes
- [ ] CI passes after rebase

### CODE REVIEW

File: src/django/api/views/export.py
Comment (mstone121) [inline src/django/api/views/export.py:122]: `LATERAL` subqueries can reference columns from earlier in the `FROM` clause. This lets us use the date to find the latest submission/review rather than the id where the order might not be guaranteed.

More info can be found in `7.2.1.5. LATERAL Subqueries` of [7.2.1. The FROM Clause](https://www.postgresql.org/docs/12/queries-table-expressions.html#QUERIES-FROM).
 
```suggestion
            INNER JOIN api_utility u ON u.id = b.utility_id
            LEFT JOIN LATERAL (
                SELECT * FROM api_submission
                WHERE boundary_id = b.id
                ORDER BY created_at DESC
                LIMIT 1
            ) s ON TRUE
            LEFT JOIN api_user du ON s.created_by_id = du.id
            LEFT JOIN api_user su ON s.submitted_by_id = su.id
            LEFT JOIN LATERAL (
                SELECT * FROM api_review
                WHERE submission_id = s.id
                ORDER BY created_at DESC
                LIMIT 1
            ) r ON TRUE
            LEFT JOIN api_user ru ON r.reviewed_by_id = ru.id
            LEFT JOIN LATERAL (
                SELECT * FROM api_approval
                WHERE submission_id = s.id
                ORDER BY approved_at DESC
                LIMIT 1
            ) a ON TRUE
            LEFT JOIN api_user au ON a.approved_by_id = au.id
```

**Review by mstone121 — APPROVED** [review]
Looks good! Everything is working as expected.

It's unfortunate that Django does not have more support for database views (perhaps due to vendor differences) or else we wouldn't have to replicate status/latest logic.

File: src/django/api/views/export.py
Comment (rajadain) [reply src/django/api/views/export.py:78]: Good point. I made the all boundaries endpoint afterwards, will include them in the approved one too, and generally align the data structures.

Comment (rajadain) [reply src/django/api/views/export.py:82]: [4a6815c](https://github.com/azavea/iow-boundary-tool/pull/218/commits/4a6815cb98a0a61c66afcf62efa7b176dc262829)

File: src/django/api/views/export.py
Comment (rajadain) [reply src/django/api/views/export.py]: Ah interesting! I hadn't heard of this before. Still don't quite have an intuition after reading up on it. I think of subqueries are row-level operations, and joins as table-level operations, so it's strange to have a subquery in a join statement. I timed them in psql though and the LATERAL JOIN appears to be running faster (???) which does not intuit well.

Nevertheless, the results are solid, and the reasoning makes sense, so I'm going with it.

[ed267d5](https://github.com/azavea/iow-boundary-tool/pull/218/commits/ed267d581e48ea8ff465e25a6fba7f5096d4dca8)

**Comment by rajadain** [comment]
Thanks for taking a look and your suggestions! I was able to solve #184 after all, updated the testing instructions to reflect that.

I also reverted the commit fixing #172 since that is now fixed by this new work. Will end up dropping that before merging.

**Comment by rajadain** [comment]
Would appreciate another quick look for the later additions to this PR

File: src/app/src/components/UtilityGuard.js
Comment (mstone121) [inline src/app/src/components/UtilityGuard.js:42]:
```suggestion
    const boundaryId = pathname.match(/^\/submissions\/(\d+)/)?.[1];
```

**Review by mstone121 — APPROVED** [review]
👍 Thanks for adding the utility auto-select!

File: src/app/src/components/UtilityGuard.js
Comment (rajadain) [reply src/app/src/components/UtilityGuard.js]: Perfect. Fixed in [b555cbc](https://github.com/azavea/iow-boundary-tool/pull/218/commits/b555cbc95cb4f6234bde07de95fc5e8bf577a613)

**Comment by rajadain** [comment]
Rebased on latest `develop`, will merge once green

File: src/app/src/components/UtilityGuard.js
Comment (rajadain) [inline src/app/src/components/UtilityGuard.js:32]: We use this here instead of `useBoundaryId` because this component is above the `Route` that declares that parameter

**Comment by rajadain** [comment]
Thanks for the multiple thorough reviews and all the great suggestions!

**Pull request merged**
